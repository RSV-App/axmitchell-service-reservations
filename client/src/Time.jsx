import React from 'react';
import PropTypes from 'prop-types';
import dateFns from 'date-fns';

class Time extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      time: '6:00 PM',
      current: new Date(),
    };

    this.handleTimeChange = this.handleTimeChange.bind(this);
    this.onChangeFuncs = this.onChangeFuncs.bind(this);
  }

  handleTimeChange(e) {
    const timeSplit = (e.target.value).split(':');
    let convertTime;
    if (timeSplit[0] === '00') {
      convertTime = '12:' + timeSplit[1];
      this.props.handleTime(convertTime);
      this.setState({
        time: convertTime,
      });
    } else if (Number(timeSplit[0]) > 12) {
      let hour = Math.abs(12 - Number(timeSplit[0])).toString();
      convertTime = hour + ':' + timeSplit[1];
      this.props.handleTime(convertTime);
      this.setState({
        time: convertTime,
      });
    } else {
      this.props.handleTime(e.target.value);
      this.setState({
        time: e.target.value,
      });
    }
  }

  onChangeFuncs(e) {
    this.handleTimeChange(e);
    this.props.getBtnBack();
  }

  render() {
    const times = [];
    const currentHour = dateFns.format(new Date(), 'H');
    const currentMin = dateFns.format(new Date(), 'm');

    if (dateFns.format(this.props.selectedDate, 'M/D/YY') === dateFns.format(new Date(), 'M/D/YY')) {
      for (let i = currentHour; i < 24; i++) {
        if (i === 0) {
          times.push(
            <option key={i} value={`${i}0:00 AM`}>
              {i + 12}:00 AM
            </option>,
            <option key={`${i}:30`} value={`${i}0:30 AM`}>
              {i + 12}:30 AM
            </option>
          );
        } else if (i > 0) {
          if (i > 12) {
            times.push(
              <option key={i} value={`${i}:00 PM`}>
                {i - 12}:00 PM
              </option>,
              <option key={`${i}:30`} value={`${i}:30 PM`}>
                {i - 12}:30 PM
              </option>
            );
          } else if (i < 12) {
            times.push(
              <option key={i} value={`${i}:00 AM`}>
                {i}:00 AM
              </option>,
              <option key={`${i}:30`} value={`${i}:30 AM`}>
                {i}:30 AM
              </option>
            );
          } else if (i == 12) {
            times.push(
              <option key={i} value={`${i}:00 PM`}>
                {i}:00 PM
              </option>,
              <option key={`${i}:30`} value={`${i}:30 PM`}>
                {i}:30 PM
              </option>
            );
          }
        }
      }
      if (currentMin <= 30) {
        times.splice(0, 1);
      }
      if (currentMin > 30) {
        times.splice(0, 2);
      }
    } else {
      for (let i = 0; i < 24; i++) {
        if (i === 0) {
          times.push(
            <option key={i} value={`${i}0:00 AM`}>
              {i + 12}:00 AM
						</option>,
            <option key={`${i}:30`} value={`${i}0:30 AM`}>
              {i + 12}:30 AM
						</option>
          );
        } else if (i > 0) {
          if (i > 12) {
            times.push(
              <option key={i} value={`${i}:00 PM`}>
                {i - 12}:00 PM
							</option>,
              <option key={`${i}:30`} value={`${i}:30 PM`}>
                {i - 12}:30 PM
							</option>
            );
          } else if (i < 12) {
            times.push(
              <option key={i} value={`${i}:00 AM`}>
                {i}:00 AM
							</option>,
              <option key={`${i}:30`} value={`${i}:30 AM`}>
                {i}:30 AM
							</option>
            );
          } else if (i === 12) {
            times.push(
              <option key={i} value={`${i}:00 PM`}>
                {i}:00 PM
							</option>,
              <option key={`${i}:30`} value={`${i}:30 PM`}>
                {i}:30 PM
							</option>
            );
          }
        }
      }
    }

    return (
      <div className="time-wrapper">
        <div className="time-title">Time</div>
        <select className="time-list" defaultValue="6:00 PM" onChange={(e) => { this.onChangeFuncs(e); }}>
          {times}
        </select>
      </div>
    );
  }
}

Time.propTypes = {
  handleTime: PropTypes.func,
  getBtnBack: PropTypes.func,
  selectedDate: PropTypes.any.isRequired,
};

export default Time;
