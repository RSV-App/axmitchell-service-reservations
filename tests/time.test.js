import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import Time from '../client/src/Time.jsx';

configure({ adapter: new Adapter() });

it('renders correctly', () => {
  const tree = renderer
    .create(<Time />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});


describe('Time List Rendering', () => {
  const wrapper = shallow(<Time />);

  it('renders a select drop down list', () => {
    const list = wrapper.find('.time-list');
    expect(list.length).toBe(1);
    expect(list.find('.time-selected').length).toBe(48);
  });

  it('should have a default time set to 6 PM', () => {
    expect(wrapper.state('time')).toBe('6:00 PM');
  });
});
