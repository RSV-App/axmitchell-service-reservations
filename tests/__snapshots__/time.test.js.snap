// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="time-wrapper"
>
  <div
    className="time-title"
  >
    Time
  </div>
  <select
    className="time-list"
    defaultValue="6:00 PM"
    onChange={[Function]}
  >
    <option
      className="time-selected"
      value="12:00 AM"
    >
      12:00 AM
    </option>
    <option
      className="time-selected"
      value="12:30 AM"
    >
      12:30 AM
    </option>
    <option
      className="time-selected"
      value="1:00 AM"
    >
      1:00 AM
    </option>
    <option
      className="time-selected"
      value="1:30 AM"
    >
      1:30 AM
    </option>
    <option
      className="time-selected"
      value="2:00 AM"
    >
      2:00 AM
    </option>
    <option
      className="time-selected"
      value="2:30 AM"
    >
      2:30 AM
    </option>
    <option
      className="time-selected"
      value="3:00 AM"
    >
      3:00 AM
    </option>
    <option
      className="time-selected"
      value="3:30 AM"
    >
      3:30 AM
    </option>
    <option
      className="time-selected"
      value="4:00 AM"
    >
      4:00 AM
    </option>
    <option
      className="time-selected"
      value="4:30 AM"
    >
      4:30 AM
    </option>
    <option
      className="time-selected"
      value="5:00 AM"
    >
      5:00 AM
    </option>
    <option
      className="time-selected"
      value="5:30 AM"
    >
      5:30 AM
    </option>
    <option
      className="time-selected"
      value="6:00 AM"
    >
      6:00 AM
    </option>
    <option
      className="time-selected"
      value="6:30 AM"
    >
      6:30 AM
    </option>
    <option
      className="time-selected"
      value="7:00 AM"
    >
      7:00 AM
    </option>
    <option
      className="time-selected"
      value="7:30 AM"
    >
      7:30 AM
    </option>
    <option
      className="time-selected"
      value="8:00 AM"
    >
      8:00 AM
    </option>
    <option
      className="time-selected"
      value="8:30 AM"
    >
      8:30 AM
    </option>
    <option
      className="time-selected"
      value="9:00 AM"
    >
      9:00 AM
    </option>
    <option
      className="time-selected"
      value="9:30 AM"
    >
      9:30 AM
    </option>
    <option
      className="time-selected"
      value="10:00 AM"
    >
      10:00 AM
    </option>
    <option
      className="time-selected"
      value="10:30 AM"
    >
      10:30 AM
    </option>
    <option
      className="time-selected"
      value="11:00 AM"
    >
      11:00 AM
    </option>
    <option
      className="time-selected"
      value="11:30 AM"
    >
      11:30 AM
    </option>
    <option
      className="time-selected"
      value="12:00 PM"
    >
      12:00 PM
    </option>
    <option
      className="time-selected"
      value="12:30 PM"
    >
      12:30 PM
    </option>
    <option
      className="time-selected"
      value="1:00 PM"
    >
      1:00 PM
    </option>
    <option
      className="time-selected"
      value="1:30 PM"
    >
      1:30 PM
    </option>
    <option
      className="time-selected"
      value="2:00 PM"
    >
      2:00 PM
    </option>
    <option
      className="time-selected"
      value="2:30 PM"
    >
      2:30 PM
    </option>
    <option
      className="time-selected"
      value="3:00 PM"
    >
      3:00 PM
    </option>
    <option
      className="time-selected"
      value="3:30 PM"
    >
      3:30 PM
    </option>
    <option
      className="time-selected"
      value="4:00 PM"
    >
      4:00 PM
    </option>
    <option
      className="time-selected"
      value="4:30 PM"
    >
      4:30 PM
    </option>
    <option
      className="time-selected"
      value="5:00 PM"
    >
      5:00 PM
    </option>
    <option
      className="time-selected"
      value="5:30 PM"
    >
      5:30 PM
    </option>
    <option
      className="time-selected"
      value="6:00 PM"
    >
      6:00 PM
    </option>
    <option
      className="time-selected"
      value="6:30 PM"
    >
      6:30 PM
    </option>
    <option
      className="time-selected"
      value="7:00 PM"
    >
      7:00 PM
    </option>
    <option
      className="time-selected"
      value="7:30 PM"
    >
      7:30 PM
    </option>
    <option
      className="time-selected"
      value="8:00 PM"
    >
      8:00 PM
    </option>
    <option
      className="time-selected"
      value="8:30 PM"
    >
      8:30 PM
    </option>
    <option
      className="time-selected"
      value="9:00 PM"
    >
      9:00 PM
    </option>
    <option
      className="time-selected"
      value="9:30 PM"
    >
      9:30 PM
    </option>
    <option
      className="time-selected"
      value="10:00 PM"
    >
      10:00 PM
    </option>
    <option
      className="time-selected"
      value="10:30 PM"
    >
      10:30 PM
    </option>
    <option
      className="time-selected"
      value="11:00 PM"
    >
      11:00 PM
    </option>
    <option
      className="time-selected"
      value="11:30 PM"
    >
      11:30 PM
    </option>
  </select>
</div>
`;
